File: src/main/java/ting/leetcode/java/linkedlist/LinkedListCycle.java
package ting.leetcode.java.linkedlist;

public class LinkedListCycle {
    public boolean hasCycle(ListNode head) {
        ListNode firstPointer = head;
        ListNode secondPointer = head;

        while (firstPointer != null) {
            firstPointer = firstPointer.next;
            if (firstPointer != null) {
                firstPointer = firstPointer.next;
                secondPointer = secondPointer.next;
            }

            if (firstPointer == secondPointer) {
                return true;
            }
        }

        return false;
    }
}

File: src/main/kotlin/ting/leetcode/linkedlist/LinkedListCycle.kt
package ting.leetcode.linkedlist

class LinkedListCycle {
    fun hasCycle(head: ListNode?): Boolean {
        var slowPointer = head
        var fastPointer = head

        while (slowPointer?.next != null) {
            slowPointer = slowPointer.next
            fastPointer = fastPointer?.next?.next ?: break
            if (slowPointer == fastPointer) return true
        }
        return false
    }
}

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

